src/code/sys_matrix.c

```diff
@@ -293,12 +293,12 @@ void Matrix_RotateZ(f32 z, u8 mode) {
 }
 
 /**
- * Rotates the top of the matrix stack by `z` degrees, then
- * rotates that matrix by `y` degrees, then rotates that matrix
- * by `x` degrees. (roll-pitch-yaw)
+ * Rotate using ZYX Tait-Bryan angles.
+ * This means a (column) vector is first rotated around X, then around Y, then around Z, then (if `mode` is
+ * `MTXMODE_APPLY`) gets transformed according to whatever the matrix was before adding the ZYX rotation.
  * Original Name: Matrix_RotateXYZ, changed to reflect rotation order.
  */
```

Comment:
what order is backwards? `ZYX` or "first rotated around X, then around Y, then around Z" ? in the sentence I'm reading the multiplication with a vector right-to-left

Reply:
Sorry for the ambiguity -- I think it should be called `ZYX` and the comment should say "first rotated around Z, then around Y, then around X." 

If it were to be decomposed into 3 axis-aligned rotation matrices, you're right it would be `M_x * M_y * M_z`. But, the most common convention I've seen is that the axes are in the left-to-right order they are applied. 

I think this is also relevant to your comment about the functions with `Translate`: I think `RotateZYXTranslate` is the best name, because the translate happens *after* the rotations.

Reply:
But with a `ZYX` rotation matrix `rotZ * rotY * rotX` when multiplying a column vector like `rotZ * rotY * rotX * vec` the vector is indeed rotated "first around X then Y then Z" ?

I also think it would be better to use `RotateZYXTranslate` so that the transform can be read left to right, but since the rotation "ZYX" is to be read right-to-left anyway I think it can also be confusing

Reply:
I think I figured out why our perspectives aren't aligned: I think it's a difference of intrinsic vs. extrinsic rotations. 
To convert between intrinsic and extrinsic rotations, reverse the order of the angles. This means "extrinsic `XYZ`" is the _same_ transformation as "intrinsic `ZYX`". 

`rotZ * rotY * rotX * vec` is the *extrinsic* rotation "first around X, then Y, then Z". However, if you perform the *intrinsic* rotation "first around Z, then Y, then X", you can derive the same value (as done in the [NASA pdf, pages 5-7](https://ntrs.nasa.gov/api/citations/19770024290/downloads/19770024290.pdf) with a diagram)

```
out = rotZ * vec''
vec'' = rotY * vec'
vec' = rotX * vec
out = rotZ * rotY * rotX * vec
```

Intrinsic rotations seem to be more common in non-graphics engineering contexts, ([e.g. aircraft](https://en.wikipedia.org/wiki/Aircraft_principal_axes), robotics). But, according to [these lecture slides, page 12](https://www.cs.utexas.edu/~theshark/courses/cs354/lectures/cs354-14.pdf), extrinsic rotations are more common in (modern-day) computer graphics. I think this is where we had a disconnect.

I think there is also an error on the Wikipedia page. There's a quote in the _[Rotation Matrix](https://en.wikipedia.org/wiki/Euler_angles#Rotation_matrix)_ section that implies that the table labels are for extrinsic rotations:
> 3. Each matrix is meant to represent, primarily, the composition of extrinsic rotations (which corresponds to the constructive evaluation of the R matrix by the multiplication of three truly elemental matrices) and, secondarily, the composition of three non elemental matrices representing globally an intrinsic rotations (around the axes of the rotating reference frame, in inverse order).

However, this line was [heavily edited after the table was added](https://en.wikipedia.org/w/index.php?title=Euler_angles&diff=782505400&oldid=765621715) -- the original text said that the table was for intrinsic rotations! :facepalm: 

---

So from here, I think we should be explicit in the function's comment that this is for intrinsic ZYX angles. 

I don't mean to cloud the discussion, but I also would be fine switching to the extrinsic convention which seems to be more popular in computer graphics. (So this function would be renamed back to `Matrix_RotateXYZ`, I guess matching the "original name," wherever that came from?)

Overall, sorry for the confusion. The nomenclature is not my strong suit, and I didn't realize that extrinsic rotations were commonly used. 
Although I like using [NASA pdf](https://ntrs.nasa.gov/api/citations/19770024290/downloads/19770024290.pdf) as a reference: it is concise, relatively complete, publicly available, and authoritative -- but it's weakness is that it doesn't cover extrinsic rotations, so maybe it was my mistake to try to introduce it here. (I do still uphold my opinion that yaw/pitch/roll should *not* be used, though :grin:)

Reply:
> I guess matching the "original name," wherever that came from?
lmao, I think original in this case is actually referring to what we called it very early on, before it was renamed to RPY. So someone a while back was accidentally misleading. I cant find evidence that it was actually called that from any debug prints

Reply:
Yeah, big :facepalm: on my part for adding to the confusion...

I'm trying to fix this error in the Wikipedia article (although we'll see if it gets reverted like Dragorn's edits did!)

Reply:
The "original name" being `Matrix_RotateXYZ` comes from the Animal Crossing symbol map